source/code/plugins/agent_maintenance_script.rb

```diff
@@ -171,32 +188,105 @@ def apply_certificate_update_endpoint(server_resp, check_for_renew_request = tru
 
     # Update the DSC_ENDPOINT variable in omsadmin.conf from the server XML
     def apply_dsc_endpoint(server_resp)
-      dsc_endpoint = ""
+      @pending_DSC_ENDPOINT = ""
 
       # Extract the DSC endpoint from the server response
       dsc_conf_regex = /<DscConfiguration.*<Endpoint>(?<endpoint>.*)<\/Endpoint>.*DscConfiguration>/
       dsc_conf_regex.match(server_resp) { |match|
-        dsc_endpoint = match["endpoint"]
+        @pending_DSC_ENDPOINT = match["endpoint"]
         # Insert escape characters before open and closed parentheses
-        dsc_endpoint = dsc_endpoint.gsub("(", "\\\\(").gsub(")", "\\\\)")
+        @pending_DSC_ENDPOINT = @pending_DSC_ENDPOINT.gsub("(", "\\\\(").gsub(")", "\\\\)")
       }
 
-      if dsc_endpoint.empty?
+      if @pending_DSC_ENDPOINT.empty?
         log_error("Could not extract the DSC endpoint.")
         return OMS::ERROR_EXTRACTING_ATTRIBUTES
       end
 
       # Update omsadmin.conf with dsc_endpoint variable
       # When apply_dsc_endpoint is called from onboarding, dsc_endpoint will be returned in file
-      update_config("DSC_ENDPOINT", dsc_endpoint)
+      if @current_DSC_ENDPOINT != @pending_DSC_ENDPOINT
+        log_info("DSC endpoint change detected, old: #{@current_DSC_ENDPOINT}, new: #{@pending_DSC_ENDPOINT}. Updating omsadmin.conf ...")
+        update_config("DSC_ENDPOINT", @pending_DSC_ENDPOINT)
+        @current_DSC_ENDPOINT = @pending_DSC_ENDPOINT
+        @pending_metaconfig_generation = true
+      end
+
+      return @current_DSC_ENDPOINT
+    end
+
+    # Generate the DSC MetaConfig file. If the dsc endpoint is updated, DSC will
+    # not directly pick the endpoint from omsadmin.conf (instead, it is picked
+    # from MetaConfig.mof), so we must manually trigger metaconfig (re)generation.
+    def generate_dsc_metaconfig
+      # Avoid calling the generation script if unnecessary
+      return if !@pending_metaconfig_generation
+
+      # If DSC is disabled, we delay MetaConfig generation until the next topology request where it has been re-enabled
+      if OMS::Configuration.is_dsc_disabled
+        log_info("DSC has been disabled, skipping DSC metaconfig generation")
+        return
+      end
+
+      retries = 0
+      max_retries = 3
+      log_info("Beginning DSC megaconfig generation")
+      while retries < max_retries
+        success = call_oms_metaconfighelper
+        if success
+          @pending_metaconfig_generation = false
+          return
+        else
+          sleep 5
+          retries += 1
+        end
+      end
+
+      log_warn("Exceeded max attempts to generate DSC metaconfig, will retry after next topology request")
```

Comment:
If all the retries are exhausted, we are saying that the dsc endpoint is updated in next topology request. 

In the first request, we call the update_config with new endpoint which i assume will update the omsadmin.conf. If all the retries fail we never revert this change, how will we able to identify the change in endpoint in this scenario for next topology requests ?

Reply:
Think of it in terms of invariants

1. omsadmin.conf will always be updated with the newest DSC_ENDPOINT following every topology request
2. pending_metaconfig_generation will always be set when a change is detected *and never unset until the metaconfig is successfully generated*.

Since the flag is never unset there is no reason to revert the change to omsadmin.conf; when the next topology request comes around metaconfig generation will be retried.

Reply:
Would it mean the state of variable 'pending_metaconfig_generation' is always persisted. Even for different executions of the script. I am assuming that this is not executed by a continuous process and there is an initialization of the variable to false in the start. Sorry if it is obvious, I did not get how it works in rb.

Reply:
No problem. It is persisted because this file defines a class `Maintenance` which is instantiated in `in_heartbeat_request.rb`, which is an input plugin that starts up and runs for the entire life cycle of the agent. You can see in that file we call `@maintenance_script.heartbeat` (do a topology request) and `@maintenance_script.generate_dsc_metaconfig` (call this method).

Reply:
Wouldn't machine restart effect this ?